ledger/store/src/helpers/rocksdb/internal/nested_map.rs

```diff
@@ -0,0 +1,1890 @@
+        Ok(raw_map_key)
+    }
+
+    // #[inline]
+    // fn get_map_raw(&self, map: &M) -> Result<Option<rocksdb::DBPinnableSlice>> {
+    //     let raw_map = self.create_prefixed_map(map)?;
+    //     match self.database.get_pinned(&raw_map)? {
+    //         Some(data) => Ok(Some(data)),
+    //         None => Ok(None),
+    //     }
+    // }
+
+    #[inline]
+    fn get_map_key_raw(&self, map: &M, key: &K) -> Result<Option<rocksdb::DBPinnableSlice>> {
+        let raw_map_key = self.create_prefixed_map_key(map, key)?;
+        match self.database.get_pinned(&raw_map_key)? {
+            Some(data) => Ok(Some(data)),
+            None => Ok(None),
+        }
+    }
+}
+
+impl<
+    'a,
+    M: 'a + Copy + Clone + Debug + PartialEq + Eq + Hash + Serialize + DeserializeOwned + Send + Sync,
+    K: 'a + Clone + Debug + PartialEq + Eq + Serialize + DeserializeOwned + Send + Sync,
+    V: 'a + Clone + PartialEq + Eq + Serialize + DeserializeOwned + Send + Sync,
+> NestedMap<'a, M, K, V> for NestedDataMap<M, K, V>
+{
+    ///
+    /// Inserts the given map-key-value pair.
+    ///
+    fn insert(&self, map: M, key: K, value: V) -> Result<()> {
+        // Determine if an atomic batch is in progress.
+        match self.is_atomic_in_progress() {
+            // If a batch is in progress, add the map-key-value pair to the batch.
+            true => self.atomic_batch.lock().push((map, Some(key), Some(value))),
+            // Otherwise, insert the key-value pair directly into the map.
+            false => {
+                // Prepare the prefixed map-key and serialized value.
+                let raw_key = self.create_prefixed_map_key(&map, &key)?;
+                let raw_value = bincode::serialize(&value)?;
+                self.database.put(raw_key, raw_value)?;
+            }
+        }
+        Ok(())
+    }
+
+    ///
+    /// Removes the given map.
+    ///
+    fn remove_map(&self, map: &M) -> Result<()> {
+        // Determine if an atomic batch is in progress.
+        match self.is_atomic_in_progress() {
+            // If a batch is in progress, add the map-None pair to the batch.
+            true => self.atomic_batch.lock().push((*map, None, None)),
+            // Otherwise, remove the map directly from the map.
+            false => {
+                let map = self.create_prefixed_map(map)?;
+                let mut map_with_separator = map.clone();
+                map_with_separator.extend(SEPARATOR);
+
+                // Iterate over the keys with the specified map prefix.
+                let entries_to_delete: Vec<_> = self
+                    .database
+                    .iterator(rocksdb::IteratorMode::From(&map, rocksdb::Direction::Forward))
+                    .filter_map(|entry| {
+                        let (map_key, _) = entry.ok()?;
+                        let map_key = map_key.to_vec();
+                        if map_key.starts_with(&map) && map_key != map_with_separator { Some(map_key) } else { None }
```

Comment:
```suggestion
                        if map_key.starts_with(&map) && map_key != map_with_separator { Some(map_key.clone()) } else { None }
```
so that we can avoid clones in the filtered-out cases

Reply:
actually this has become quite long; it's probably better that I push a commit